### ISSUE TITLE
rust-analyzer incorrectly reports every mention of `Self` as a reference to an associated type

### ISSUE THREAD

**Issue by SludgePhD** [rust-lang/rust-analyzer#17803]
**rust-analyzer version**:`rust-analyzer version: 0.3.2053-standalone [/home/sludge/.vscode/extensions/rust-lang.rust-analyzer-0.3.2053-linux-x64/server/rust-analyzer]`

**rustc version**: `rustc 1.80.0 (051478957 2024-07-21)`

**code snippet to reproduce**: 
```rust
pub trait TestTrait {
    type Assoc;
    fn stuff() -> Self;
}
impl TestTrait for () {
    type Assoc = u8;
    fn stuff() -> Self {
        let me: Self = ();
        me
    }
}
```

When the cursor is on one of the `Assoc` tokens, every instance of `Self` in the trait impl is considered a mention of the associated type:

![screenshot_24-08-05_04:57:02](https://github.com/user-attachments/assets/f04caff4-6d13-42ae-af1e-d356475d7d23)

This seems wrong (and it also happens when using find references), since `Assoc` is not in fact the same as `()` in that impl.

**Comment by roife** [comment]
Goto-references mistakenly treats `Self` as a reference of `Assoc`.
